### PR TITLE
let lasers bounce with zero energy for one tick

### LINKED ISSUE #5380: Laser does no longer reflect from walls on map Putt Putt
I will try to bisect the commit. It is already broken since 16.0.3

### PR DESCRIPTION
**Pull request by C0D3D3V**
fixes #5380 

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] FIXES physics to old physics
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by C0D3D3V** [comment]
One thing to note here. I do not know if you somehow could trigger two zero energy bounces in a row in the old physics. I tried my best to trigger it somehow, but it seems to be only possible if you are stuck in a wall.

File: src/game/server/entities/laser.cpp [hunk 1]
```diff
@@ -150,12 +151,20 @@ void CLaser::DoBounce()
...
+			}
 			else if(!m_TuneZone)
+			{
 				m_Energy -= Distance + GameServer()->Tuning()->m_LaserBounceCost;
+				m_ZeroEnergyBounceInLastTick = Distance == 0.0f;
+			}
 			else
+			{
 				m_Energy -= distance(m_From, m_Pos) + GameServer()->TuningList()[m_TuneZone].m_LaserBounceCost;
+				m_ZeroEnergyBounceInLastTick = Distance == 0.0f;
```
Comment (def-) [inline src/game/server/entities/laser.cpp:166]: Could move this line to after the if-elseif-else, since it is always done.

**Review by def- — CHANGES_REQUESTED** [review]
Otherwise fine. Thanks, let's hope this is the last change required here.

**Comment by C0D3D3V** [comment]
I also finished Putt Putt with this change :D everything seems to be smooth :)

**Review by def- — APPROVED** [review]

File: src/game/server/entities/laser.h [hunk 2]
```diff
@@ -30,6 +30,7 @@ class CLaser : public CEntity
 	int m_EvalTick;
 	int m_Owner;
 	int m_TeamMask;
+	int m_ZeroEnergyBounceInLastTick;
```
Comment (Chairn) [inline src/game/server/entities/laser.h:33]: bool?

File: src/game/client/prediction/entities/laser.h [hunk 3]
```diff
@@ -34,6 +34,7 @@ class CLaser : public CEntity
 	int m_Bounces;
 	int m_EvalTick;
 	int m_Owner;
+	int m_ZeroEnergyBounceInLastTick;
```
Comment (Chairn) [inline src/game/client/prediction/entities/laser.h:37]: bool?

Comment (C0D3D3V) [reply src/game/client/prediction/entities/laser.h:37]: mh I will fix it in an other PR...

Comment (C0D3D3V) [reply src/game/client/prediction/entities/laser.h:37]: I experimented first if there is a need to allow more then one zero energy bounce, but then turned it later into a bool :D but forgot somhow to change it in the definition.

**Pull request merged**